src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -251,6 +254,18 @@ public bool Restore(SolutionRestoreRequest request)
                         // Initialize if not already done.
                         await InitializeAsync();
 
+#if VS15
+                        if (request.RestoreSource == RestoreOperationSource.OnBuild)
+                        {
+                            if (!await SolutionManager.SolutionHasDeferredProjectsAsync()
+                                && SolutionManager.GetNuGetProjects().All(p => p is CpsPackageReferenceProject))
+                            {
```

Comment:
The `RestoreManagerPackage` has nothing to do with restore logic. Its role is to re-route UI events to the worker. The `SolutionRestoreWorker` in its turn owns this decision. It helps keeping all conditionals in one place accessible via the same entry point.

Also I'd argue on `JTF` usage here as every query to `SolutionManager` runs under `JTF` anyway.